NRF_51_DK - adapt lp/us ticker drivers to the new standards

mprse opened the pull request:
## Description

1. us ticker driver for NRF_51_DK board:
According to new ticker standards the following requirements for us ticker are not met on NRF_5xxx boards:
- has a frequency between 250KHz and 8MHz (currently is driven by 32kHz clock)
- ticker increments by 1 each tick (currently is scaled to 1 MHz by incrementing counter by ~31)

Since BLE softdevice uses TIMER0 the proposition is to use high speed TIMER1 for us ticker configured as follows:
- TIMER counter width: 16 bits (max)
- TIMER frequency: 1MHz
This solution also uses Timer's capture/compare register 0 to specify interrupt time and Timer's capture/compare register 1 to read current timer value.

<del>
2. sleep/deep-sleep driver for NRF_5xxx boards:
According to new sleep mode standards the following requirement is not met on NRF_5xxx boards:
- High-speed clocks are turned off in deep-sleep mode.
Additionally when ticker resolution is 31 us, then it is hard to verify that wake-up time does not exceeds 10 us (fix for us ticker solve this problem). 

The proposition is to  stop timer used by us ticker before going to deep-sleep and enable it after wake-up. This operation should result in disabling Timer clock sources in deep-sleep mode.

According to documentation:
If the system does not require any of the clocks provided by the HFCLK clock controller, the HFCLK
controller may enter a power saving mode automatically and switch off the selected clock source. This
occurs if all peripherals that require either PCLK1M, PCLK16M are appropriately stopped or disabled, and
the CPU is sleeping and thereby no longer requesting HCLK. 

</del>

3. lp ticker driver for NRF51_DK board:
- According to NRF51_DK reference manual rtc interrupt cannot be controlled by rtc event. In the previous implementation interrupts were enabled permanently and specific interrupt was enabled/disabled by enabling/disabling the specific event. If event is enabled, then event signal is provided to Programmable Peripheral Interconnect (PPI). If interrupt is enabled, then interrupt signal is provided to Nested Vector Interrupt Controller (NVIC). Disable all events permanently. Enable lp ticker overflow interrupt permanently(needed for RTC), disable lp ticker capture/compare interrupt on init (lp_ticker_init) , enable lp ticker interrupt when lp ticker interrupt is set (lp_ticker_set_interrupt), disable lp ticker interrupt on disable request(lp_ticker_disable_interrupt).
- Provide lp ticker data for higher level (freq: 32kHz / len: 24 bits),
- Add the following features to init routine: disable lp ticker interrupt.
- Make ticker driver to operate on ticks instead of us.
- Simplify lp ticker read and set interrupt routines (upper layers handles conversion to us and interrupt scheduling).

## Status

**IN DEVELOPMENT**

## Migrations

YES

API is not changed, but the behavior of us ticker and deep-sleep mode changes as follows:
- us ticker is driven by Timer 1 incremented by 1 with frequency 1MHz/16 bits,
- Timer's Capture/Compare register 0 is used to set interrupt time,
- Timer's Capture/Compare register 1 is used to get current timer count
- us ticker init routine disables us ticker interrupt,
- lp ticker driver operates on ticks,
- lp ticker init routine disables lp ticker interrupt
- lp ticker is not scaled to 1MHz

bulislaw commented:
CC: @c1728p9 @0xc0170

bulislaw commented:
@mprse that should probably go to the branch with HAL changes
@pan- any hints how to conform this board to our new HAL APIs without hurting power?

pan- commented:
@bulislaw I think the situation could be better if the event queue used a Timer object with an lp_ticker rather than the a us_ticker (note that this trap is not documented in the header, the default constructor documentation is empty).

bulislaw commented:
Is the accuracy of LP good enough?

pan- commented:
The event queue works at the ms scale (1kHz); not the us scale. According to the specification lp tickers should have a reported frequency between 8KHz and 64KHz  (see [here](https://github.com/ARMmbed/mbed-os/blob/204fe4459e8f41b3fdf613ad58182b1603a9d40e/hal/lp_ticker_api.h#L37)). 

So yes, the lp ticker should be enough for the event queue. 
@c1728p9 @geky Could you confirm ?

geky commented:
👍 Yep, 1ms resolution is all that's needed, we just haven't defaulted to LowPowerTickers since it's not clearly defined what resolution garuntees there are.

mprse commented:
Power consumption tests were performed on NRF51_DK board (battery powered). Board has been modified according to the NRF51_DK user guide (chapter 6.7.1 - [link](http://infocenter.nordicsemi.com/pdf/nRF51_Development_Kit_User_Guide_v1.2.pdf)) as follows:
> 1. Cut the PCB track shorting solder bridge SB9 to put P22 in series with the load.
> 2. Short solder bridge SB11 (if using coin cell battery) or SB12 (if using external power supply) to
> bypass the protection diode which would otherwise give a voltage drop.

Test scenario:
```
- init lp, us tickers
- wait - 10 sec
- sleep - 10 sec
- deep-sleep - 10 sec
```

Test results:

  | branch: master | branch: nrf5x_use_high_speed_timer_for_us_timer
-- | -- | --
Normal operation | 4 [uA] | 558 [uA]
Sleep mode | 4 [uA] | 558 [uA]
Deep-sleep mode | 4 [uA] | 4 [uA]

I have additionally checked power consumption before us ticker init routine is called and it was equal to 558 [uA]. I'm a little confused, how this is possible (us ticker should be disabled before init)?

pan- commented:
@mprse Thank you for the analysis, that is really helpful. Could you do the power analysis with a BLE application like the [BLE beacon example](https://github.com/ARMmbed/mbed-os-example-ble/blob/master/BLE_Beacon/source/main.cpp) ? To avoid initialization of the UART peripheral, please [remove line 84](https://github.com/ARMmbed/mbed-os-example-ble/blob/master/BLE_Beacon/source/main.cpp#L84).

mprse commented:
> @mprse Thank you for the analysis, that is really helpful. Could you do the power analysis with a BLE application like the BLE beacon example ? To avoid initialization of the UART peripheral, please remove line 84.


  | branch: master | branch: nrf5x_use_high_speed_timer_for_us_timer
-- | -- | --
BLE beacon example (without line 84) | 3.5 [mA] | 4.4 [mA]

mprse commented:
I found that typical capacity of CR2032 battery is 240 mAh, so `BLE beacon example` can be run on battery power:
- 64,5 h when us ticker uses low power timer (master),
- 53,5 h when us ticker uses high speed timer (nrf5x_use_high_speed_timer_for_us_timer).

So the difference is 11 h (performance decreased by 17%).

pan- commented:
@mprse I'm a bit surprised by the current results; did you compile with the release profile or develop profile ?

mprse commented:
> @mprse I'm a bit surprised by the current results; did you compile with the release profile or develop profile ?

With the default profile:
`mbed test -t GCC_ARM -m NRF51_DK -n tests-mbed_hal-ble --compile -v -c`

bulislaw commented:
@pan- what do you think?

pan- commented:
@bulislaw @mprse  I'm really surprised by the numbers. On The NRF52; if I disable the trace that print the MAC address and compile with the release profile then the power consumption of the Beacon example is very low with master: ~20uA.

The Beacon example doesn't work with that PR however.

mprse commented:
@pan- I noticed that the Beacon example doesn't work on NRF51 on this branch either, so the provided power consumption tests are invalid.


The Beacon example generates exception during BLE initialisation in the following [line](https://github.com/ARMmbed/mbed-os/blob/master/features/FEATURE_BLE/targets/TARGET_NORDIC/TARGET_NRF5/source/btle/btle.cpp#L132). I observed that `sd_softdevice_enable` system call returns `4097` error code which does not tell me much. It looks like these changes have influence on BLE drivers. Maybe someone who is familiar with BLE driver can provide some hint?

pan- commented:
@mprse The error return is `NRF_ERROR_SDM_INCORRECT_INTERRUPT_CONFIGURATION`; it means that the  softsevice interrupt is already enabled, or an enabled interrupt has an illegal priority level.

In our case it is an illegal priority level of the us_ticker IRQ. Please use the lowest priority for the us ticker interrupt (see [lp_ticker](https://github.com/mprse/mbed-os/blob/1c96df6ce9aaf587f5174c42d1e33eb5d4b64fb0/targets/TARGET_NORDIC/TARGET_NRF5/common_rtc.c#L166-L172)).

mprse commented:
@pan- Thanks for the provided explanation of the problem. I have provided a fix to us_ticker_init routine in order to enable TIMER0 irq with the lowest priority. 
Unfortunately I still get the same error in the Beacon example program.

mprse commented:
It looks like enabling TIMER0 irq (`nrf_drv_common_irq_enable(TIMER0_IRQn, <priority>)`) regardless of irq priority which is requested causes that BLE initialisation (`sd_softdevice_enable()`) fails with the `NRF_ERROR_SDM_INCORRECT_INTERRUPT_CONFIGURATION` error code. If TIMER0 interrupt is disabled, then BLE initialisation is successful.

pan- commented:
@mprse Looking at the product specification of the softdevice, TIMER0 is used by the softdevice and shouldn't be used by application code while the softdevice is active. TIMER1 and TIMER2 are free however. My comment on priority stands.

mprse commented:
@pan- Thanks. This was the problem. I used TIMER1 instead of TIMER0 and Beacon example runs now on NRF51_DK board. Unfortunately TIMER1 and TIMER2 are 8 or 16 bits wide, so I had to slightly modify us ticker driver. In current version us ticker has 16 bits and its frequency is 250 KHz.

Checked that HAL tests for for us/lp ticker and sleep passes on this branch.

mprse commented:
Current measurement results on NRF51_DK:


  | wait() | hal_sleep() | hal_deepsleep() | BLE beacon example (without line 84)
-- | -- | -- | -- | --
master | ~4.3 [mA] | ~4 [uA] | ~3 [uA] | ~4.5 [mA]
this branch | ~4.5 [mA] | ~555 [uA] | ~4 [uA] | ~4.5 [mA]

bulislaw commented:
I think that's fair enough, @pan- what do you think?

bulislaw commented:
That should be on the new ticker hal api branch I think.

pan- commented:
> That should be on the new ticker hal api branch I think.

+1.

mprse commented:
> 
> 
>     That should be on the new ticker hal api branch I think.
> 
> +1.
@pan @bulislaw 
I can rebase this PR and put it on feature-hal-spec-ticker branch, but please note that this PR provides also fix for feature-hal-spec-sleep branch.

bulislaw commented:
Ah my point was it should go to one (or more if needed) as the feature is not yet on master and won't be for some time.

mprse commented:
Changed destination branch from master to ARMmbed:feature-hal-spec-ticker. Rebased   mprse:nrf5x_use_high_speed_timer_for_us_timer on ARMmbed:feature-hal-spec-ticker.

cmonr commented:
@pan- Does this still need a review?

adbridge commented:
@cmonr @pan- Looks like a re-review is needed.

0xc0170 commented:
I'll run quick CI check as it has not been for some time, while we review

/morph build

0xc0170 commented:
@mprse Please verify the latest test failure, looks related

mprse commented:
> @mprse Please verify the latest test failure, looks related

It looks like there is some problem with low power ticker. Under investigation.

mprse commented:
Since provided changes have huge impact on the higher level tests results the ticker drivers needs to be updated in two steps (1 - update of us ticker driver, 2 - update of lp ticker driver). Removed last commit (lp ticker driver update) and slightly modified driver for us ticker (set timer clock to 1MHz - this is the closest solution to the previous version). When this pass CI (us ticker driver is ok) I will provide update of lp ticker driver.
Note to pass CI tests to new ticker HAL API are required provided in PR https://github.com/ARMmbed/mbed-os/pull/5233.

mprse commented:
Re-based after test to the new ticker HAL API have been merged to the destination branch and enabled support for NRF51_DK board.

bulislaw commented:
@pan- please re-review.

mprse commented:
> @pan- please re-review.

Please wait until is ready - investigating issue with lp ticker. It looks like it is related to issue https://github.com/ARMmbed/mbed-os/issues/5835.

mprse commented:
Re-based after destination branch has been re-based to master (which includes patches: #5889 and #5884).

0xc0170 commented:
/morph build

mprse commented:
> Build : FAILURE

Provided fix for NRF52_DK (conditional compilation issue).

Unfortunately it looks like new ticker hal layer (with synchronisation of public functions) is not compatible with target specific ticker drivers provided here. We can run morph to get full picture.

0xc0170 commented:
/morph build

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5/sleep.c:
```diff
@@ -71,10 +77,20 @@ void hal_sleep(void)
     }
+
+    if (deepsleep) {
+        nrf_timer_task_trigger(NRF_TIMER1, NRF_TIMER_TASK_START);
```
This should be conditional of the timer state before deepsleep entry. There is no reason to start the peripheral after deep sleep if it wasn't started earlier.

mprse replied:
If `MBED_TICKLESS` is enabled, then `usticker` is enabled (`RtosTimer`). It looks like all NRF5 boards have tick-less mode enabled, so `usticker` is always in use.

pan- replied:
I know I've commented about this issue earlier and I've made a mistake. `RtosTimer` use the lp_ticker, not the us_ticker: 

https://github.com/ARMmbed/mbed-os/blob/f1cf77fa443ee48b29861a8581f5d3f685c7ec2e/rtos/TARGET_CORTEX/mbed_rtx_idle.cpp#L50-L51

In this code `_ticker_data` is initialized by the `TimerEvent` constructor with the lp ticker. 

------

Considering defects or current implementation details in separate modules to build new code is a good way to build a fragile house of cards. It is expected  `hal_deepsleep` will restore us_ticker operating state at wakeup time so we should do just that. If the us ticker was started before deepsleep then start it again otherwise; do nothing.

Cleaning incorrect uses of the us ticker (event queue, `wait_ms`) will be a mandatory before this branch gets into master. Better to not add work we will have to do eventually.

mprse replied:
I this case I'll provide a fix to enable `usticker` only if it was enabled before going to sleep.

mprse replied:
Fixed.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5/sleep.c:
```diff
@@ -18,15 +18,21 @@
 #define FPU_EXCEPTION_MASK 0x0000009F
 
-void hal_sleep(void)
+static void sleep_common(bool deepsleep)
```
I'm not sure that it's necessary to factor things in sleep_common. The conditional bits introduced when `deepsleep` is true are a prolog and epilog of the `hal_sleep` function. Why not have those in `hal_deepsleep`  and remove the conditions ?

mprse replied:
So you suggest to move `sleep_common` code to `hal_sleep` and implement `hal_deepsleep` as follows:
```
- disable usticker
- call hall_sleep()
- enable usticker
```

I think it is a good fix which will increase code readability.

mprse replied:
Fixed.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5/sleep.c:
```diff
@@ -75,6 +77,15 @@ void hal_sleep(void)
 {
+    if (us_ticker_is_enabled()) {
+        /* Disable timer used for us ticker. */
+        nrf_timer_task_trigger(NRF_TIMER1, NRF_TIMER_TASK_STOP);
```
Disabling the ticker should update the variable accessed with `us_ticker_is_enabled`; why not add a function that takes care of that ? 
This state should also be stored so it can be used after wake up.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5/sleep.c:
```diff
@@ -75,6 +77,15 @@ void hal_sleep(void)
     hal_sleep();
     //   NRF_POWER->SYSTEMOFF=1;
+
+    if (us_ticker_is_enabled()) {
```
Please acquire locally the state of the ticker at the function entry then use this state to decide if the ticker should be enable or not. A sane implementation of `us_ticker_is_enabled` would always return false at this point.

pan- commented on targets/TARGET_NORDIC/TARGET_NRF5/sleep.c:
```diff
@@ -75,6 +77,15 @@ void hal_sleep(void)
     //   NRF_POWER->SYSTEMOFF=1;
+
+    if (us_ticker_is_enabled()) {
+        nrf_timer_task_trigger(NRF_TIMER1, NRF_TIMER_TASK_START);
```
Maybe a dedicated function can takes care of that and update `us_ticker_enabled` at the same time.

mprse replied:
`us_ticker_enabled` variable informs if ticker has been initialised (enabled). Enabled ticker can not be disabled, because there is no API to disable the ticker. 
I assume that during the deep-sleep ticker is still enabled, but its counter is frozen. In case when ticker is enabled on wake-up we need to revive its counter.

pan- replied:
I understood the logic before making my comments however having a hidden undocumented function named `us_ticker_is_enabled` returning true when the us_ticker is at best misleading.

If you want to keep the logic, please rename the function `us_ticker_is_enabled` into something less ambiguous. `us_ticker_is_initialized` may be a relevant name . 

Note that soon there will be API to disable tickers: https://github.com/ARMmbed/mbed-os/pull/5856 .

marcuschangarm commented:
Hi @mprse - I'm getting an interesting error when running this simple application compiled with the debug profile:

```
#include "mbed.h"

DigitalOut led1(LED1, 0);

int main(void)
{
    printf("hello world\r\n");

    for (;;)
    {
        printf("blink!\r\n");

        led1 = !led1;
        wait_ms(1000);
    }
}
```

Output:
```
hello world
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
blink!
mbed assertation failed: os_timer->get_tick() == svcRtxKernelGetTickCount(), file: ./mbed-os/rtos/TARGET_CORTEX/mbed_rtx_idle.c▒▒▒?
```

0xc0170 commented:
@marcuschangarm That debug error was reported few days ago, should be fixed on master (that assertation was removed)

mprse commented:
Hi @marcuschangarm,

I know there is a problem with new lp ticker driver. It looks like interrupt scheduling is broken. Interrupts are scheduled in the past and ticker irq handler is executed one by one by means of `fire_interrupt` routine. I believe that this is related to ticker frequency - in the new version target specific driver only defines frequency for upper (HAL) layer (32.786 Hz) and operates on ticks (i.e. read routine returns number of ticks without scaling to us; HAL layer is responsible to convert ticks to us and vice versa). So the additional operations need to be performed, but I can not find what is wrong. 

I have the following test which shows the issue:
```
volatile int cnt = 0;
void isr()
{
    cnt++;
}

void test()
{
   Timer timer;
   LowPowerTicker ticker;

   timer.start();
   ticker.attach_us(isr, 1000); // 1 ms period

   while(timer.read() < 1);

   ticker.detach();
   timer.stop();
   
   printf("---> cnt: %d \n", cnt); // ~3000
   printf("---> cnt: %f \n", timer.read()); // ~3sec
}
```
It shows that main thread does not get CPU time and exit from while loop is performed at least 2 seconds later. If I change defined frequency of the lp ticker to 1000000Hz, then it looks like results are positive taking into account that real clock speed is 32768 Hz.

bulislaw commented:
What's the status of this PR? What's left to be done?

bulislaw commented:
@jamesbeyond

cmonr commented:
@bulislaw It looks like the PRs and issues that were previously referenced have recently made it into master. Next steps would be a rebase and a local retest before CI is run again.

cmonr commented:
@pan- should also re-review changes since the last review is a tad bit stale.

mprse commented:
> What's the status of this PR? What's left to be done?

Mainly its ready. But this change has impact on higher layer ticker tests. It looks like it is related to defined ticker frequency. When I define lp ticker frequency 32768Hz and remove scaling to 1MHz by the driver (HAL layer shell handle ticks<--->us conversion) then there is a problem with interrupt scheduling. 

I have suspended this for now and I'm checking if this issue also exists on other boards (K64F).

mprse commented:
@c1728p9 Since you are an author of the new Ticker HAL API I thought that maybe you could review this PR? Could you check if the NRF51_DK driver implementation is consistent with the new HAL Ticker API?

cmonr commented:
@mprse  Rebase this PR when you get the chance, as there are changes in master that are needed for the littlefs travis check.

Or is this PR still suspended?

mprse commented:
> @mprse Rebase this PR when you get the chance, as there are changes in master that are needed for the littlefs travis check.
> 
> Or is this PR still suspended?

I noticed some CI problems which looks not related to this PR. I think ticker feature branch needs to be re-based on master, but when the following fixes are on master:
- PR https://github.com/ARMmbed/mbed-os/pull/6068 (Fix for issue #6054 - interrupts scheduled in the past),
- PR https://github.com/ARMmbed/mbed-os/pull/6076 (Remove ticker interface initialization while Ticker creation).

cmonr commented:
@mprse I just updated and rebased the feature branch. When you get a chance, go ahead and rebase this PR, and fingers crossed this solves our issues.

mprse commented:
Waiting for the following patches:
- PR https://github.com/ARMmbed/mbed-os/pull/6076,
- PR https://github.com/ARMmbed/mbed-os/pull/6068,
- PR https://github.com/ARMmbed/mbed-os/pull/6052 (lp/us ticker test update).

mprse commented:
> Waiting for the following patches:
> 
> PR #6076,
> PR #6068,
> PR #6052 (lp/us ticker test update).

All listed PR have been merged. Rebased this PR on ticker feature branch after K64F ticker port with test updates have been merged (https://github.com/ARMmbed/mbed-os/commit/5440cc4155a21c7f7adea4bd83e596d19465e85f). 
Can we run morph here?

0xc0170 commented:
/morph build

cmonr commented:
@mprse Heads up, it looks like this will need to be rebased again before it's merged.

mprse commented:
> Test : FAILURE
> 
> Build number : 1072
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/5629/1072

Updated this PR on the destination branch and resolve conflicts.
The destination branch has been rebased while morph build. I checked locally on GCC ARM and all tests have passed on NRF51_DK, so I think morph needs to be repeated, but first Jenkins/Travis issues needs to be resolved.

mprse commented:
@cmonr @adbridge Can we run morph build here?

adbridge commented:
@pan- Can you please re-review this and approve if happy...

adbridge commented:
/morph build

mprse commented:
Failing tests analysis:
- `tests-mbed_drivers-lp_ticker`: Time measurement constrains are not met. Can not reproduce locally. 
- `tests-mbed_drivers-timer`:  Time measurement constrains are not met. Can not reproduce locally. 
- `tests-mbed_drivers-timerevent`: This is a new test. It should not be executed if USTICKER support is disabled.  Additionally `TimerEvent` class test is based on us ticker and does not provides protection against entering into deep-sleep mode when delay is performed. In current implementation us ticker is disabled in deep-sleep mode - this will be probably changed since this feature should be enabled on sleep feature branch.
- `tests-events-timing`: This test occasionally fails. In some cases `1 ms` delay performed in loop using `wait_ms` function takes about ~28 ms. It looks like the reason of this additional time is entering into sleep mode. This issue also exists without disabling the high speed counter before entering into the deep-sleep mode https://github.com/ARMmbed/mbed-os/pull/5629/commits/80f542f57724541087a4e051bfbe573119858e93.

Next step:
- `tests-mbed_drivers-timerevent` perform test only if  USTICKER support is enabled.
- remove sleep fix from this PR.
- disable sleep support for now, to check if sleep is the reason of all failures.

mprse commented:
Can we run morph here? @adbridge @cmonr

adbridge commented:
@mprse This needs the review signing off before we put it into CI. @donatieng who could review this in @pan- absence this week ?

cmonr commented:
@donatieng If you're not the right person to review, we're open to suggestions.

0xc0170 commented:
/morph build

mprse commented:
> @mprse This needs the review signing off before we put it into CI.

@adbridge  I understand that review needs to be done before CI, but please note that two reviewers have approved the changes already. I know that in the meantime I have done some changes, but these changes were related to the CI failures. Additionally we need to have ticker enabled on two platforms as soon as possible. I do not have vpn access and can use only GCC_ARM compiler so morph results are necessary for me so I can go forward with this PR.

donatieng commented on TESTS/mbed_hal/lp_us_tickers/main.cpp:
```diff
@@ -65,7 +67,10 @@ uint32_t count_ticks(uint32_t cycles, uint32_t step)
 
-    return (stop - start);
+    /* Handle overflow - overflow protection may not work in this case. */
+    uint32_t diff = (start <= stop) ? (stop - start) : (uint32_t)(max_count - start + stop);
```
Should be
```C
uint32_t diff = (start <= stop) ? (stop - start) : (uint32_t)(max_count + 1 - start + stop);
```

mprse replied:
Fixed.

donatieng commented on TESTS/mbed_hal/lp_us_tickers/main.cpp:
```diff
@@ -302,9 +307,23 @@ void ticker_increment_test(void)
+                 * to repeat counting with the same number of cycles.
+                 * In cases if difference is exactly 1 we can exit the loop.
+                 */
+                if (next_tick_count - base_tick_count == 1 ||
```
Isn't that what this test is validating in the first place? If the core is too slow, you could reduce NUM_OF_CYCLES instead.

mprse replied:
Fixed.

donatieng commented on TESTS/mbed_hal/lp_us_tickers/main.cpp:
```diff
@@ -406,7 +427,7 @@ utest::v1::status_t hf_ticker_setup(const Case *const source, const size_t index
     ticker_overflow_delta = HF_TICKER_OVERFLOW_DELTA;
 
-    overflow_protect();
+    overflow_protect(20000);
```
Please use macro

mprse replied:
Fixed.

mprse commented:
@donatieng Thanks for the review. I'll provide fixes soon.

> In terms of power measurements what did you use?

Power consumption tests were performed on NRF51_DK board (battery powered). Board has been modified according to the NRF51_DK user guide (chapter 6.7.1 - link) as follows:

> Cut the PCB track shorting solder bridge SB9 to put P22 in series with the load.
> Short solder bridge SB11 (if using coin cell battery) or SB12 (if using external power supply) to
> bypass the protection diode which would otherwise give a voltage drop.

> It seems that the beacon example was tested with debug & UART enabled? 4.5mA seems a lot.

I think that serial port and debug prints were removed from the beacon example. Note that in current form this PR provides only example implementation of us/lp ticker driver for vendors which is consistent with the new standards where target specific driver function are simplified and ticker HAL layer is handling conversion between counter ticks and us based on configured counter freq and time tracing based on counter size. Unfortunately I have still some issues.

mprse commented:
`tests-mbed_drivers-lp_ticker`: test fails on CI while testing `LowPowerTicker.attach()` function with timestamp set to 500 ms:
`[1520435617.20][CONN][RXD] :170::FAIL: Values Not Within Delta 2000 Expected 500000 Was 506037`

I have created the following test locally to check the timing issue:

```
void test(void)
{
    LowPowerTicker ticker;

    int delays[] = {1000, 5000, 10000, 20000, 30000, 40000, 50000, 60000, 70000, 80000, 90000, 100000, 500000, 1000000};
    int times[(sizeof(delays)/sizeof(int))];
    int now[(sizeof(delays)/sizeof(int))];

    uint32_t int_set_buf[100];
    uint32_t curr_time_buf[100];

    for(int i = 0; i < (sizeof(delays)/sizeof(int)); i++) {
        ticker_callback_flag = 0;
        gtimer.reset();
        gtimer.start();

        ticker.attach_us(callback(stop_gtimer_set_flag), delays[i]);
        while(!ticker_callback_flag);
        ticker.detach();
        times[i] = gtimer.read_us();

        int_set_buf[i] = int_set; // tick count written to cc register
        curr_time_buf[i] = curr_time; // current tick count while writing to cc register
    }

    for (int i = 0; i < (sizeof(times)/sizeof(int)); i++) {
            printf("---> delay: %7d; m_time: %7d m_diff: %4d set_int_tick: %7d sched_tick: %7d diff: %7d \n",
                    delays[i], times[i], times[i] - delays[i], curr_time_buf[i], int_set_buf[i], int_set_buf[i] - curr_time_buf[i]);
    }
}
```
The aim of this test is to check how much time elapses between attaching low power ticker interrupt and interrupt generation time for different timeout values. The Timer used to measure the time is stopped in the interrupt handler. For testing purposes I have disabled `SYS_TICK`, so lp ticker is only used by this test and run the test on master and this branch with the following results:

```
this branch:
delay:    1000; m_time:    1141 m_diff:   141 set_int_tick:       8 sched_tick:      37 diff:      29
delay:    5000; m_time:    5135 m_diff:   135 set_int_tick:      51 sched_tick:     212 diff:     161
delay:   10000; m_time:   10116 m_diff:   116 set_int_tick:     227 sched_tick:     551 diff:     324
delay:   20000; m_time:   20088 m_diff:    88 set_int_tick:     566 sched_tick:    1218 diff:     652
delay:   30000; m_time:   30087 m_diff:    87 set_int_tick:    1233 sched_tick:    2214 diff:     981
delay:   40000; m_time:   40030 m_diff:    30 set_int_tick:    2229 sched_tick:    3537 diff:    1308
delay:   50000; m_time:   50002 m_diff:     2 set_int_tick:    3552 sched_tick:    5188 diff:    1636
delay:   60000; m_time:   59975 m_diff:   -25 set_int_tick:    5203 sched_tick:    7167 diff:    1964
delay:   70000; m_time:   69916 m_diff:   -84 set_int_tick:    7182 sched_tick:    9473 diff:    2291
delay:   80000; m_time:   79889 m_diff:  -111 set_int_tick:    9488 sched_tick:   12107 diff:    2619
delay:   90000; m_time:   89860 m_diff:  -140 set_int_tick:   12122 sched_tick:   15069 diff:    2947
delay:  100000; m_time:   99795 m_diff:  -205 set_int_tick:   15084 sched_tick:   18358 diff:    3274
delay:  500000; m_time:  498284 m_diff: -1716 set_int_tick:   18373 sched_tick:   34755 diff:   16382
delay: 1000000; m_time:  996431 m_diff: -3569 set_int_tick:   34769 sched_tick:   67534 diff:   32765

master:
delay:    1000; m_time:    1252 m_diff:  252 set_int_tick:   38915 sched_tick:   38944 diff:      29
delay:    5000; m_time:    5249 m_diff:  249 set_int_tick:   38969 sched_tick:   39129 diff:     160
delay:   10000; m_time:   10254 m_diff:  254 set_int_tick:   39155 sched_tick:   39478 diff:     323
delay:   20000; m_time:   20263 m_diff:  263 set_int_tick:   39504 sched_tick:   40155 diff:     651
delay:   30000; m_time:   30274 m_diff:  274 set_int_tick:   40180 sched_tick:   41160 diff:     980
delay:   40000; m_time:   40253 m_diff:  253 set_int_tick:   41185 sched_tick:   42492 diff:    1307
delay:   50000; m_time:   50262 m_diff:  262 set_int_tick:   42517 sched_tick:   44152 diff:    1635
delay:   60000; m_time:   60272 m_diff:  272 set_int_tick:   44178 sched_tick:   46140 diff:    1962
delay:   70000; m_time:   70251 m_diff:  251 set_int_tick:   46166 sched_tick:   48455 diff:    2289
delay:   80000; m_time:   80261 m_diff:  261 set_int_tick:   48480 sched_tick:   51098 diff:    2618
delay:   90000; m_time:   90271 m_diff:  271 set_int_tick:   51124 sched_tick:   54069 diff:    2945
delay:  100000; m_time:  100250 m_diff:  250 set_int_tick:   54095 sched_tick:   57367 diff:    3272
delay:  500000; m_time:  500274 m_diff:  274 set_int_tick:   57392 sched_tick:   73773 diff:   16381
delay: 1000000; m_time: 1000244 m_diff:  244 set_int_tick:   73798 sched_tick:  106562 diff:   32764

delay - LowPowerTicker attach delay value.
m_time - measured time using Timer between attach() call and interrupt handler execution.
m_diff - difference between delay and measured time.
set_int_tick - current RTC counter value while setting the interrupt by means of lp_ticker_set_interrupt().
sched_tick - tick count written to RTC capture compare register while setting interrupt.
diff - difference between current tick and tick when interrupt is to be generated.

```
Results analysis:
 - On master measurement error is nearly constant while on this branch the errors grows together with increasing the LowPowerTimer timeout value. 
- The interrupt scheduling (values written to capture/compare register) is consistent on both branches.

The difference between this two branches:
- On this branch us ticker, which is used for measuring the elapsed time, is based on high freq 1 MHz 16 bit timer while on master is based on RTC (same as lp ticker). 
- On this branch HAL Ticker layer is responsible for conversion between ticks and us, while on master NRF ticker driver handles conversion between ticks and us.


Conclusions:

- Test results proves that HAL Ticker layer properly handles interrupt scheduling and us<--->ticks conversion since values written to RTC cc register on master and this branch are nearly the same.
- On master the same hardware counter is used to measure elapsed time and generate lp ticker interrupts, so this is why the measurement error is constant. On this branch 1 MHz higher precision clock has been used to measure the elapsed time and that is why the results are different. The results shows that RTC clock is inaccurate and measurement error grows (in this case grows down - interrupts are generated faster than they should be). I have checked also that operations between starting the timer and writing value to RTC cc register takes about ~3 ticks which corresponds to 120 ms. This explains why for small delays measured time is greater than assumed.

The problem is that the results are different on CI and when I run `tests-mbed_drivers-lp_ticker` test locally. For the same test case (500 ms delay) on CI the measured time is ~505 ms and locally I get ~498.5 ms. I checked that the test is build using the same flags. The only solution taking into account above analysis is that RTC on CI boards is also inaccurate, but it is slightly slower than 32.768 Hz while on my board RTC clock is slightly faster. 

Looking into the results on CI, the RTC accuracy is about 10% which is no so bad. Always for more precise measurement us ticker can be used. The test needs to be updated to and hard-coded delta value equal to 2000 us needs to be changed for NRF51_DK board.

mprse commented:
`tests-mbed_drivers-timer` : On this branch us ticker is based on 1 MHz higher precision counter (on master is based on 32kHz RTC). As a result of using higher precision counter measurement error increased:
:485::FAIL: Expected 1.060000 Was 1.070478

For delay 1060 ms we measured ~1070 ms so we have about ~1% measurement error which is not bed for this board.

To make it work with 1MHz us ticker test tolerance needs to be increased.

mprse commented:
@donatieng Fixed after review.
Analyzed recent results and provided fixes for the failed tests.

@0xc0170 Can we run morph?

0xc0170 commented:
/morph build

mprse commented:
> Test : FAILURE
> 
> Build number : 1190
> Test logs :http://mbed-os-logs.s3-website-us-west-1.amazonaws.com/?prefix=logs/5629/1190

`tests-mbed_drivers-timer`: last test case is failing 1s time measurement:
:735::FAIL: Expected 1.000000 Was 1.008932

Measurement error is less than 1%. The test needs to be updated.

mprse commented:
`tests-mbed_drivers-timer`: changed implementation of the tolerance factor to be dependent on delay time:
`delta = 500 us + 0.02 * delay_us` 

@0xc0170 Can we run morph?

0xc0170 commented:
/morph build

cmonr commented:
Test CI canceled due to 5.8.0-RC2 generation. Will resume once 5.8.0-RC2 is released.

cmonr commented:
A different build was turned on and retriggered builds. Stopping them for RC2.

mprse commented:
> Test CI canceled due to 5.8.0-RC2 generation. Will resume once 5.8.0-RC2 is released.
@0xc0170 @cmonr 
Can we resume morph?

0xc0170 commented:
Will do soon, the last CI check is running for rc2 at the moment

mprse commented:
@0xc0170 ping

0xc0170 commented:
/morph build

cmonr commented:
Going to need approval from @pan- and @donatieng, even once this passes CI.

mprse commented:
I think this one is ready for now. I still need to add some modifications, but this will be done in the separate PR. All changes requested by @pan @donatieng  are processed. Can you check please?

donatieng commented on TESTS/mbed_drivers/timer/main.cpp:
```diff
@@ -30,35 +30,24 @@ using namespace utest::v1;
+ * 10 ms delay: tolerance = 700 us
+ * 100 ms delay: tolerance = 2500 us
+ * 1000 ms delay: tolerance = 20500 us
+ *
```
This new way of defining tolerance seems like a pretty big change from the previous model, will have to check with the relevant people in the team

donatieng commented:
@c1728p9 If you could have a look at the tolerance changes?

mprse commented:
Since NRF51_DK was the worst board for timing tests (slow and with both tickers based on 32KHz RTC clock), the tolerance factor was adjusted to this board. Now when the us ticker is based on 1 MHz counter it shows the inaccuracy of the RTC clock and tolerance in timing tests has to be adjusted again. 
@c1728p9 can you take a look on this?

bulislaw commented:
@c1728p9 @mprse ping

mprse commented:
Rebased this PR. But it looks like there are some problems with jenkins.

cmonr commented:
@mprse We restarted both the pr-head and travis jobs. Travis failed due to being unable to download packages successfully, and we've been working on resolving an issue with pr-head since yesterday.

Also, a couple of other PRs came into the branch, which means this will need another rebase.

mprse commented:
Rebased this PR and removed updates of hal ticker tests which have been moved to PR https://github.com/ARMmbed/mbed-os/pull/6472.

0xc0170 commented:
@mprse Please rebase and time for another review round !

mprse commented:
> @mprse Please rebase and time for another review round !

Rebased!

0xc0170 commented:
> Thanks for addressing my review. One outstanding point is a major change in the definition of tolerance in timer tests which will need to be checked by relevant team members.

Has this been resolved?

mprse commented:
> > Thanks for addressing my review. One outstanding point is a major change in the definition of tolerance in timer tests which will need to be checked by relevant team members.
> 
> Has this been resolved?

No, someone needs to review this. Now we have 2% tolerance for us ticker and 5% tolerance for lp ticker instead of constant value.

jamesbeyond commented:
could you review this please @donatieng @c1728p9

mprse commented:
Re-based this PR after feature branch has been re-based.

mprse commented:
Added fix for lp ticker driver and re-enabled sleep support for NRF51_DK.

mprse commented:
@0xc0170 Can we run morph after last updates?

0xc0170 commented:
>@0xc0170 Can we run morph after last updates?

Will label this as needs: CI (still howeve waiting for all request changes to approved!). At the moment , release candidate is under test (+ one fix that needs to get in there).

cmonr commented:
/morph build

cmonr commented:
Going on a hunch that a Jenkins machine left too soon.

Will leave for @0xc0170 to confirm and restart, since the build artifacts contain more than just the usual PASS directory.

0xc0170 commented:
/morph export-build

cmonr commented:
@pan- @donatieng Are y'all happy with this PR?

c1728p9 commented:
The tolerance definitions look good to me.
